File: IT/Eccezioni/Esercizi/C++/esercizio25.cpp
/*
file esercizio25.cpp
compilare con: g++ -std=c++11 esercizio25.cpp
*/

#include <iostream>
#include <array>
#include <stdexcept>

using std::cout;
using std::endl;
using std::cerr;
using std::string;
using std::to_string;
using std::ostream;
using std::logic_error;

class MyDate {
	private:
		int yyyy;
		int mm;
		int dd;

	public:
		MyDate(int, int, int);
		int getDD();
		int getMM();
		int getYYYY();
		void setDD(int);
		void setMM(int);
		void setYYYY(int);

	friend ostream& operator<<(ostream&, const MyDate&);
};

MyDate::MyDate(int dd, int mm, int yyyy) {
	try {
		this->setYYYY(yyyy);
		this->setMM(mm);
		this->setDD(dd);
	}
	catch (logic_error& e) {
		cerr << e.what() << endl;
		this->setYYYY(1970);
		this->setMM(1);
		this->setDD(1);
	}
}

int MyDate::getDD() {
	return this->dd;
}

void MyDate::setDD(int dd) {
	if (dd < 1) {
		throw logic_error("Giorno (" + to_string(dd) + ") non valido!");
	}

	int minDD = 1;
	int maxDD = 0;

	if (this->mm == 4 || this->mm == 6 || this->mm == 9 || this->mm == 11) {
		maxDD = 30;
	}
	else if (this->mm == 2) {
		maxDD = 28;
	}
	else {
		maxDD = 31;
	}

	if (dd > maxDD) {
		throw logic_error("Giorno (" + to_string(dd) + ") non valido!");
	}

	this->dd = dd;
}

int MyDate::getMM() {
	return this->mm;
}

void MyDate::setMM(int mm) {
	if (mm < 1 || mm > 12) {
		throw logic_error("Mese (" + to_string(mm) + ") non valido!");
	}
	this->mm = mm;
}

int MyDate::getYYYY() {
	return this->yyyy;
}

void MyDate::setYYYY(int yyyy) {
	if (yyyy <= 0) {
		throw logic_error("Anno (" + to_string(yyyy) + ") non valido!");
	}
	this->yyyy = yyyy;
}

ostream& operator<<(ostream& os, const MyDate& d) {
	return os << d.dd << "-" << d.mm << "-" << d.yyyy;
}

int main()
{
	MyDate d(1, 1, 2000);
	cout << d << endl;

	d = MyDate(1, 13, 2000);
	cout << d << endl;
}

File: IT/Eccezioni/Esercizi/Python/esercizio25.py
# file esercizio25.py
# -*- coding: utf-8 -*-

from __future__ import print_function
if hasattr(__builtins__, 'raw_input'): input = raw_input

class MyDate(object):
	def __init__(self, dd, mm, yyyy):
		try:
			self.yyyy = yyyy
			self.mm = mm
			self.dd = dd
		except ValueError as e:
			print(e)
			self.yyyy = 1970
			self.mm = 1
			self.dd = 1

	def getDD(self):
		return self.__dd

	def setDD(self, dd):
		minDD = 1
		maxDD = 0

		if dd < 1:
			raise ValueError("Giorno ({}) non valido!".format(dd))

		if self.mm in [4, 6, 9, 11]:
			maxDD = 30
		elif self.mm == 2:
			maxDD = 28
		else:
			maxDD = 31

		if dd > maxDD:
			raise ValueError("Giorno ({}) non valido!".format(dd))

		self.__dd = dd

	def getMM(self):
		return self.__mm

	def setMM(self, mm):
		if mm < 1 or mm > 12:
			raise ValueError("Mese ({}) non valido!".format(mm))
		self.__mm = mm

	def getYYYY(self):
		return self.__yyyy

	def setYYYY(self, yyyy):
		if yyyy <= 0:
			raise ValueError("Anno ({}) non valido!".format(yyyy))
		self.__yyyy = yyyy

	def __str__(self):
		return "{}-{}-{}".format(self.dd, self.mm, self.yyyy)

	dd = property(getDD, setDD)
	mm = property(getMM, setMM)
	yyyy = property(getYYYY, setYYYY)

d = MyDate(dd=1, mm=1, yyyy=2000)
print(d)

d = MyDate(dd=1, mm=13, yyyy=2000)
print(d)

